design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,582 @@
+if the request uses the force parameter it must perform the extra authentication(s).
+
+This can also be handled on a configuration level for clients, but it is simply
+a matter of creating a specific authentication flow, with the desired number of
+authentication factors, and the required authenticator categories, and then setting
+the authentication flow to a client.
+
+For both SAML and OIDC, however, this will require being able to define respectively
+[authentication context classes](https://docs.oasis-open.org/security/saml/v2.0/saml-authn-context-2.0-os.pdf) and [authentication method reference values](https://tools.ietf.org/html/rfc8176) to an authentication category. This is necessary to be able to transmit a
+value within an `<AuthContext>` (for SAML) or an `amr` claim (for OIDC). It is also necessary
+for Keycloak to be able to present during an authentication with a `<RequestedAuthnContext>`
+the user with the correct choice of authenticators.
+
+### Modifications to the admin console
+
+#### Authentication > flows section
+
+Authentication factors must be able to be configured with an LoA, which can be
+any positive value. Subsequent authentication factors must have their values set
+to equal or increasing values. A string value may also be set as an alias for the
+LoA, to be used for example when the LoA is described as an URI. If the alias is
+not set, the numeral value is sent to the client in the assertion, and if it
+is set, the alias is used. For ease of information, the interface should show
+that the authentication factor is being used as a LoA, and at which level.
+
+#### Client
+
+In the client configuration it must be possible to set a LoA before being able to
+have access to the client. Since the LoA is attached to the flow, it should be within
+the same GUI section as the choice of authentication flow, and for coherence, it should
+be only among the choices of the flow. However, behind the scenes this only sets
+the level for the client. If the configuration of the authentication flow is changed
+and the value of the LoA modified, this does not change the value within the client. Within the
+configuration of the client, it may show that the current LoA is not relevant to the
+configured authentication flow as a visual aid.
+
+Note that for OIDC clients, this means that a client can have a different LoA
+requirements for browser and direct grant flows.
+
+### Modifications to the REST API
+
+There are only minimal modifications to the REST API, and they mirror the changes
+to the GUI:
+* Possibility to set the LoA and alias when setting or modifying an authentication factor
+* Possibility to set the LoA for the browser or direct grant flow when creating or editing a client
+
+### Extending protocol implementation within Keycloak
+
+The following protocols or optional parts of protocols must be implemented for OIDC:
+* Handling of the `acr` claim, `acr_values` parameter and `prompt` parameter as
```

Comment:
I think the `prompt` parameter is already implemented so that you can force users to log in even though they have an active session on the server. Or are you referring to something more specific?

Reply:
I confess that I don't know what is already implemented here, just that it's useful for the step-up. But if it's already implemented, all the better.

Reply:
An interesting thing here is that OIDC has prompt=login which says "SHOULD prompt the End-User for reauthentication". What reauthentication means may differ though. It may be asking for password, otp, or both. Not sure how we would handle that and make it configurable what reauthentication means.

Reply:
Currently "prompt=none" just skips the cookie. I think we can just preserve this behaviour and ensure that all authentication factors, which are marked as REQUIRED (and eventually OPTIONAL in case that user has particular credentials set) will need to be re-authenticated.

So for example, there is a flow, which has "password" authentication factor as REQUIRED and "otp" factor as OPTIONAL. Then using parameter combinations can work like this:
- "prompt-login" will make sure that user needs to re-authenticate with password. Just if he has OTP set, he will need to re-authenticate with OTP as well.

- "acr_values=otp" will ensure that user needs to authenticate (or re-authenticate) only with the OTP (No password re-authentication required). IMO it will be good if using "acr_values" will ensure that user needs to re-authenticate with OTP even if he already authenticated with it before in this session?

- "prompt=login&acr_values=otp" will ensure that user needs to authenticate (or re-authenticate) with both password as REQUIRED and OTP as REQUIRED. Password is REQUIRED as it is REQUIRED authentication factor configured and OTP is required as it is requested by the "acr_values" parameter.

Reply:
> Currently "prompt=none" just skips the cookie. I think we can just preserve this behaviour and ensure that all authentication factors, which are marked as REQUIRED (and eventually OPTIONAL in case that user has particular credentials set) will need to be re-authenticated.
> 
> So for example, there is a flow, which has "password" authentication factor as REQUIRED and "otp" factor as OPTIONAL. Then using parameter combinations can work like this:
> 
> * "prompt-login" will make sure that user needs to re-authenticate with password. Just if he has OTP set, he will need to re-authenticate with OTP as well.
> * "acr_values=otp" will ensure that user needs to authenticate (or re-authenticate) only with the OTP (No password re-authentication required). IMO it will be good if using "acr_values" will ensure that user needs to re-authenticate with OTP even if he already authenticated with it before in this session?
> * "prompt=login&acr_values=otp" will ensure that user needs to authenticate (or re-authenticate) with both password as REQUIRED and OTP as REQUIRED. Password is REQUIRED as it is REQUIRED authentication factor configured and OTP is required as it is requested by the "acr_values" parameter.

I'm not sure if is a good idea to use `acr_values` to reference authentication methods (`amr`). IMO, we should try to request re-authentication based on security levels as much as possible. Or even use a custom `amr_values` claims to indicate the expected authentication methods.

Reply:
> IMO, we should try to request re-authentication based on security levels as much as possible

I agree, this seems to be how the protocol is designed

> Or even use a custom amr_values claims to indicate the expected authentication methods.

This could be done, but as it is out of the scope of the protocol, there wouldn't be any SPs requesting this, no?

Reply:
Yes, it would be for Keycloak specific use cases if we really want to allow clients to specify a set of one or more authentication mechanisms without necessarily use the`acr_values` claims.

Reply:
I am sorry for the confusion. I didn't mean to mix `acr_values` (authentication factors/subflows) with the `amr` claims (which reference authenticators). In the example, when I used `acr_values=otp` I meant to use the authentication subflow (or authentication factor), which contains OTP authenticator, not directly the OTP authenticator itself. I should probably use some more explicit naming like `acr_values=otp-subflow` or something like that.

In other words, in my previous comment, I meant to use `prompt=login` together with specific `acr_values` to specify which authentication subflows/factors need to be re-authenticated.

IMO We don't need something like custom field `amr_values` as you can always model the flow with the `acr_values` . In other words, if you ever need to have something like request authentication with OTP, you should model subflow `otp-flow`, which will contain the OTP authenticator as the only authenticator and then reference this subflow with the acr_values claim like `acr_values=otp-flow`

Reply:
I think I understand what you mean with your examples.

> IMO it will be good if using "acr_values" will ensure that user needs to re-authenticate with OTP even if he already authenticated with it before in this session?

I'm not sure that just having `acr_values=otp_flow` should make keycloak re-authenticate with the _otp_flow_ LoA. For me, if the user is already connected with the _otp_flow_ LoA, then if the client requests it, keycloak should send over the token without asking the user for an new authentication. This is for the purpose of SSO: if client 1 and client 2 require _otp_flow_ LoA, then logging with that level to the first should automatically grant access to the 2nd.

Rather, I think that re authentication should only be done on `login=prompt`, as the protocol suggests. However, the protocol is not super clear on what should be done when `prompt=login` is used with `acr_values` (and more generally with LoA). But I think that keycloak can act exactly as is asked by the request from the client. Perhaps it's best to illustrate it with the following example:
Keycloak is configured with two authentication factor subflows, **1st factor** which provides LoA _lvl1_ and **2nd** factor which provides LoA _lvl2_. Imagining a user already logged in with a LoA _lvl2_, then
1. the client asks `prompt=login` -> keycloak re-asks the user to log in with **1st factor** subflow, as it would do on a normal login with no `acr_values` specified.
2. the client asks `prompt=login&acr_values=lvl1` --> same as case 1, but this time because it's explicitly specified.
3. the client asks `prompt=login&acr_values=lvl2` --> keycloak only executes the subflow for **2nd factor**. Note that if the user hadn't already been logged in, keycloak would also have executed the subflow for **1st factor**.

The only problem I see with this is that there's no way for a client to force keycloak to execute the subflow for the **1st factor** and for the **2nd factor**. The protocol does allow for having a list of `acr_values` (this is also true for if you're using the `acr` claim set to `essential`), but states that it's for "values appearing in order of preference". I suppose that you could have the following

4. the client asks `prompt=login&acr_value=lvl1 lvl2` --> keycloak executes the subflow for **1st factor** and **2nd factor**

But this doesn't seem to follow the specification.

Reply:
I agree with the usage of `acr_values` that you and Marek are proposing. Depending on the values and user's authentication state we perform the necessary steps to gain the requested levels.

I think what you described in #4 is covered by the spec. Why do you think it is not ? The `prompt=login` should ask user for re-authentication and `acr_values` or `acr` claim using the `claims` request parameter (alternatively) should instruct the LoA you want for a user.

Reply:
I think that the protocol doesn't clearly state what was defined in this situation. Following the definition of `acr_values` and `prompt=login` in the [OIDC protocol](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest), the declaration  `prompt=login&acr_value=lvl1 lvl2` from the client to the IdP should translate as "I want you to force the user to re-authenticate, I would prefer if he was logged in with LoA lvl1, but LoA lvl2 is OK as well". 

Making the user log in with the subflows corresonding to LoA lvl1 and LoA lvl2 is not wrong, but it's not completely correct either. But I don't see any other way in the protocol that tells the idp "first make the user authenticate with LoA lvl1, and then with LoA lvl2", aside from sending two authentication requests. If everyone is happy with the behaviour "the client asks `prompt=login&acr_value=lvl1 lvl2` --> keycloak executes the subflow for **1st factor** and **2nd factor**", then it can be implemented that way.

Reply:
Based on this statement from the spec, where ` the Authorization Server MUST return an acr Claim Value that matches one of the requested values`, I assume that if you request `prompt=login&acr_value=lvl1 lvl2`, we as a AS, are free to process whatever we find is the best as long as we return one of them. IMO, if you have N and N+1 we should always enforce the highest level. If the user is already on N we don't need to enforce N, but only N+1. If the request was for N only we do N even if the user is authenticated and set with N. If you request N+1 only we only do N+1 if user already is set with N, etc.

Does it make sense ? @mposolda , @stianst wdyt ?